File: Java/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        Map<String, String> refP = new HashMap<>();
        Map<String, Integer> sumSells = new HashMap<>();
        
        for (int i = 0; i < enroll.length; i++) {
            refP.put(enroll[i], referral[i]);
        }
        
        for (int i = 0; i < seller.length; i++) {
            String cur = seller[i];
            int money = amount[i] * 100;
            
            while (!cur.equals("-")) {
                int share = money / 10;
                int get = money - share;
                
                sumSells.put(cur, sumSells.getOrDefault(cur, 0) + get);
                
                money = share;
                cur = refP.get(cur);
                
                if (share == 0) break;
            }
        }
        
        int[] answer = new int[enroll.length];
        for (int i = 0; i < enroll.length; i++) {
            answer[i] = sumSells.getOrDefault(enroll[i], 0);
        }
        
        return answer;
    }
}

File: C++/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <cmath>

using namespace std;

// 누가 누구와 연결되어있는지 작성할 map이 필요해보임.

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    unordered_map<string, string> tree; // tree를 이용하여 누가 누구와 연결되어있는지 확인
    unordered_map<string, int> idx;
    vector<int> ans (enroll.size(), 0); // enroll의 사이즈로 0 초기화
    
    for (int i = 0; i < enroll.size(); i++) {
        idx[enroll[i]] = i;
    }
    
    for (int i = 0; i < referral.size(); i++) {
        tree.insert({enroll[i], referral[i]});
    }
    
    for (int i = 0; i < seller.size(); i++) {
        int money = amount[i] * 100; // 판매 수익
        string next = seller[i];
        
        while (next != "-") {
            int give = money / 10;
            ans[idx[next]] += money - give;
            
            if (give < 1) break;
            
            money = give;
            next = tree[next];
        }
    }
    
    return ans;
}